### PR TITLE
Create a model hierarchy inspector

### PR DESCRIPTION
**Pull request by ellthompson**
image

Displays the hierarchy of the loaded glTF model in a new model panel and the total verts / primitives contained in that model. Each node in the hierarchy can also be selected to view their local transform.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by mvaligursky** [comment]
Could we display number of meshes in a node as well? (next to vertices / indices)

**Comment by ellthompson** [comment]
> Could we display number of meshes in a node as well? (next to vertices / indices)

Sure, i've added this in.

**Comment by ellthompson** [comment]
image

File: src/controls.tsx [hunk 1]
```diff
@@ -71,6 +86,37 @@ const LightingPanel = () => {
...
+    const modelHierarchy: Array<HierarchyNode> = useObserverState(observer, 'model.nodes', true);
+    const enabled: boolean =  modelHierarchy.length > 0;
+    const mapNodes = (nodes: Array<HierarchyNode>) => {
+        return nodes.map((node:HierarchyNode) => <TreeViewItem text={`${node.name}`} key={node.path} onSelected={() => observer.set('model.selectedNode.path', node.path)}>
+            { mapNodes(node.children) }
+        </TreeViewItem>);
+    };
+    return (
+        <Panel headerText='MODEL' collapsible >
+            <Detail name='meshCount' label='Meshs:' path='model.meshCount'/>
```
Comment (mvaligursky) [inline src/controls.tsx:100]: Meshs? :)

**Comment by mvaligursky** [comment]
I believe he viewer still uses a Model component and not Render component. And that means the hierarchy matches the loaded glb, but all meshes are attached as part of a single Model on the root. Does the current solution handles displaying correct per node stats when a user selects it, or does it always display stats for the single Model?
This would not be an issue with Render component, as that closely matches the GLB.

**Comment by ellthompson** [comment]
> I believe he viewer still uses a Model component and not Render component. And that means the hierarchy matches the loaded glb, but all meshes are attached as part of a single Model on the root. Does the current solution handles displaying correct per node stats when a user selects it, or does it always display stats for the single Model?
> This would not be an issue with Render component, as that closely matches the GLB.

The per node stats are only the transform that node. The mesh / vertex / primitive counts are for the entire model.

**Comment by mvaligursky** [comment]
So that when we port the viewer to use render component (soon I guess as it's going to be public this week), we'd need to change it? Ideally those stats are per node instead of global.

Maybe we should move to render component now?

**Comment by ellthompson** [comment]
> So that when we port the viewer to use render component (soon I guess as it's going to be public this week), we'd need to change it? Ideally those stats are per node instead of global.
> 
> Maybe we should move to render component now?

This can be treated as the first pass. I'd say that can be done when we expand this into a proper inspector panel on the right with more in-depth stats. It'll require a rewrite at that point anyway.

**Comment by slimbuck** [comment]
Perhaps we could render the currently select node's axis in the viewport?

**Review by slimbuck — APPROVED** [review]

**Pull request merged**